src/world.rs

```diff
@@ -1245,6 +1246,29 @@ struct InsertTarget {
     index: u32,
 }
 
+type IndexTypeIdMap<V> = HashMap<(u32, TypeId), V, BuildHasherDefault<IndexTypeIdHasher>>;
+
+#[derive(Default)]
+struct IndexTypeIdHasher(u64);
+
+impl Hasher for IndexTypeIdHasher {
+    fn write_u32(&mut self, index: u32) {
+        self.0 = u64::from(index);
```

Comment:
Does std guarantee the order in which tuple fields are hashed? If not, this should probably be xor-assign for robustness; LLVM will probably optimize out the extra work in any case.
```suggestion
        self.0 ^= u64::from(index);
```

Reply:
> Does std guarantee the order in which tuple fields are hashed?

For derived `Hash` impls (like those for tuples), fields are always hashed in source order. See https://github.com/rust-lang/rust/pull/91660 for a recent example where this was utilised.

Reply:
(Actually, the example is not helpful as it relies on the memory layout instead of the hash order.)

Reply:
I don't follow. The linked PR is a custom `Hash` implementation that makes a single `hash` call, rendering ordering questions moot.

Reply:
My goal here is just to not rely on behavior that isn't clearly defined as part of std's public interface, at least if there's any reasonable alternative.

Reply:
I was unable to find a definite statement on the order besides the [source code](https://github.com/rust-lang/rust/blob/673d0db5e393e9c64897005b470bfeb6d5aec61b/compiler/rustc_builtin_macros/src/deriving/hash.rs#L85) itself. Changed as suggested.